### PR TITLE
Create properly-typed arrays

### LINKED ISSUE #1548: Array creation within fake call configuration fails due to type mismatch
Found while checking AutoFakeItEasy for FakeItEasy 5.0.0 compatibility.

AutoFixture.AutoFakeItEasy.UnitTest.FakeItEasyRelayTest.CreateWhenSpecificationIsSatisfiedReturnsCorrectResult happens to include

```c#
var contextStub = A.Fake<ISpecimenContext>();
A.CallTo(() => contextStub.Resolve(typeof(Fake<>).MakeGenericType(request)))
    .Returns(expected);
```
`MakeGenericType` takes a `params Type[]`, which we don't handle well:

```
AutoFixture.AutoFakeItEasy.UnitTest.FakeItEasyRelayTest.CreateWhenSpecificationIsSatisfiedReturnsCorrectResult(request: typeof(object)) [FAIL]                                                 
  System.ArgumentException : Object of type 'System.Object[]' cannot be converted to type 'System.Type[]'.                                                                                     
  Stack Trace:                                                                                                                                                                                 
       at System.RuntimeType.TryChangeType(Object value, Binder binder, CultureInfo culture, Boolean needsSpecialCast)                                                                         
       at System.Reflection.MethodBase.CheckArguments(Object[] parameters, Binder binder, BindingFlags invokeAttr, CultureInfo culture, Signature sig)                                         
       at System.Reflection.RuntimeMethodInfo.InvokeArgumentsCheck(Object obj, BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture)                               
       at System.Reflection.RuntimeMethodInfo.Invoke(Object obj, BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture)                                             
       at FakeItEasy.ExpressionExtensions.TryFastEvaluate(Expression expression, Object& result)                                                                                               
       at FakeItEasy.ExpressionExtensions.Evaluate(Expression expression)                                                                                                                      
       at FakeItEasy.Expressions.ExpressionArgumentConstraintFactory.<>c__DisplayClass14_0.<GetArgumentConstraintFromExpression>b__0()                                                         
       at FakeItEasy.Core.ArgumentConstraintTrap.TrapConstraints(Action actionThatProducesConstraint)                                                                                          
       at FakeItEasy.Expressions.ExpressionArgumentConstraintFactory.GetArgumentConstraintFromExpression(Expression expression, Type parameterType, Object& value)                             
       at FakeItEasy.Expressions.ExpressionArgumentConstraintFactory.GetArgumentConstraint(ParsedArgumentExpression argument)                                                                  
       at FakeItEasy.Expressions.ExpressionCallMatcher..ctor(ParsedCallExpression parsedExpression, ExpressionArgumentConstraintFactory constraintFactory, MethodInfoManager methodInfoManager)
       at FakeItEasy.RootModule.<>c__DisplayClass1_0.<RegisterDependencies>g__ExpressionCallRuleFactory|3(ParsedCallExpression callSpecification)                                              
       at FakeItEasy.Configuration.FakeConfigurationManager.CallTo[T](Expression`1 callSpecification)                                                                                          
    Src\AutoFakeItEasyUnitTest\FakeItEasyRelayTest.cs(133,0): at AutoFixture.AutoFakeItEasy.UnitTest.FakeItEasyRelayTest.CreateWhenSpecificationIsSatisfiedReturnsCorrectResult(Type request)  
```
`FakeItEasy.ExpressionExtensions.TryFastEvaluate` creates an `object[]` for the arguments of `MakeGenericType`, and the only element of this array is also an `object[]`. We should instead be making a `Type[]`.

Note: the `params`iness of the parameter isn't necessary. If we explicitly create an array within the call configuration expression.

### PR DESCRIPTION
**Pull request by blairconrad**
Fixes #1548.

### CODE REVIEW

File: src/FakeItEasy/ExpressionExtensions.cs [hunk 1]
```diff
@@ -123,14 +124,16 @@ private static bool TryFastEvaluate(Expression expression, out object result)
 
                 case ExpressionType.NewArrayInit:
                     var newArrayExpression = (NewArrayExpression)expression;
-                    object[] arrayItems = new object[newArrayExpression.Expressions.Count];
+                    Array arrayItems = (Array)Activator.CreateInstance(newArrayExpression.Type, newArrayExpression.Expressions.Count);
```
Comment (thomaslevesque) [inline src/FakeItEasy/ExpressionExtensions.cs]: Wouldn't `Array.CreateInstance` be faster?

File: tests/FakeItEasy.Specs/ConfigurationSpecs.cs [hunk 2]
```diff
@@ -754,6 +771,11 @@ public sealed override int ReturnSomething()
 
         public class Foo : IFoo
         {
+            public static IFoo BuildFromParamsArray(int[] integers)
```
Comment (thomaslevesque) [inline tests/FakeItEasy.Specs/ConfigurationSpecs.cs]: Why "ParamsArray"? There's no `params` modifier

File: src/FakeItEasy/ExpressionExtensions.cs (hunk 1 above)
Comment (blairconrad) [reply src/FakeItEasy/ExpressionExtensions.cs]: Maybe! I did not think about it. Will do.

Comment (blairconrad) [reply src/FakeItEasy/ExpressionExtensions.cs]: Oh, I'll have to "de-array" the type. Still worth it?

File: tests/FakeItEasy.Specs/ConfigurationSpecs.cs (hunk 2 above)
Comment (blairconrad) [reply tests/FakeItEasy.Specs/ConfigurationSpecs.cs]: Bother. I'd started with a params array and missed this when I converted.

File: src/FakeItEasy/ExpressionExtensions.cs (hunk 1 above)
Comment (thomaslevesque) [reply src/FakeItEasy/ExpressionExtensions.cs]: I did a quick benchmark, it seems to be about 13 times as fast

**Comment by blairconrad** [comment]
I amended directly. In addition to the two changes you asked for, I suppresssed CA1801 on `BuildFromArray`.

**Pull request merged**